### PR TITLE
Store and display the times when files were generated.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Looks like:

image

This will be useful as we allow users to move to a model where generators [don't run automatically](https://github.com/dotnet/roslyn/pull/72494), letting them when the generator was last run.

### CODE REVIEW

File: src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.CompilationTracker_Generators.cs [hunk 1]
```diff
@@ -325,7 +333,8 @@ private partial class CompilationTracker : ICompilationTracker
                                 generatedSource.SyntaxTree.Options,
                                 ProjectState.LanguageServices,
                                 // Compute the checksum on demand from the given source text.
-                                originalSourceTextChecksum: null));
+                                originalSourceTextChecksum: null,
+                                generationDateTime));
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.CompilationTracker_Generators.cs:337]: these are the places in the system where the generation actually happens, and so we record the time it happened as well.  From this point on, this just flows outwards to be read and stored and eventually shown in the UI.

**Comment by CyrusNajmabadi** [comment]
@genlu ptal.

**Comment by sharwell** [comment]
Oh man, do we have the ability to put a "re-run now" button in that?

**Comment by CyrusNajmabadi** [comment]
> Oh man, do we have the ability to put a "re-run now" button in that?

I def imagine we would be able to do that.

File: src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.CompilationTracker_Generators.cs [hunk 2]
```diff
@@ -281,6 +283,7 @@ private partial class CompilationTracker : ICompilationTracker
             }
 
             using var generatedDocumentsBuilder = TemporaryArray<SourceGeneratedDocumentState>.Empty;
+            var generationDateTime = DateTime.Now;
```
Comment (sharwell) [inline src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.CompilationTracker_Generators.cs:286]: 💡 Comment here to make sure it's not accidentally moved inside the `foreach` loop later. We want all files from a generation pass to have the same stamp, even though they were not all produced at exactly the same time.

Comment (CyrusNajmabadi) [reply src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.CompilationTracker_Generators.cs:286]: sure :)

**Comment by CyrusNajmabadi** [comment]
> Oh man, do we have the ability to put a "re-run now" button in that?

We can def do that.  I'll do it in the PR where we add 'balanced' mode.

**Comment by CyrusNajmabadi** [comment]
@dibarbet ptal

**Review by sharwell — APPROVED** [review]

File: src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.CompilationTracker_Generators.cs [hunk 3]
```diff
@@ -281,6 +283,9 @@ private partial class CompilationTracker : ICompilationTracker
             }
 
             using var generatedDocumentsBuilder = TemporaryArray<SourceGeneratedDocumentState>.Empty;
+
+            // Capture the date now.  We want all the generated files to use this date consistently.
+            var generationDateTime = DateTime.Now;
```
Comment (ToddGrun) [inline src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.CompilationTracker_Generators.cs:288]: >Now [](http://example.com/codeflow?start=46&length=3)

Why not UtcNow?

Comment (CyrusNajmabadi) [reply src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.CompilationTracker_Generators.cs:288]: will update in followup.

**Comment by CyrusNajmabadi** [comment]
@jasonmalinowski 	For review when you get back.

**Pull request merged**